src/sentry/types/region.py

```diff
@@ -1,8 +1,120 @@
+from __future__ import annotations
+
 from dataclasses import dataclass
+from enum import Enum
+from typing import TYPE_CHECKING, Iterable
+
+from sentry.silo import SiloMode
+
+if TYPE_CHECKING:
+    from sentry.models import Organization
+
+
+class RegionCategory(Enum):
+    GEOGRAPHIC = "GEOGRAPHIC"
+    SINGLE_TENANT = "SINGLE_TENANT"
 
 
-@dataclass
+@dataclass(frozen=True, eq=True)
 class Region:
+    """A region of the Sentry platform, hosted by a region silo."""
+
     name: str
-    subdomain: str
-    is_private: bool = False
+    """The region's unique identifier."""
+
+    id: int
+    """The region's unique numeric representation.
+
+    This number is used for composing "snowflake" IDs, and must fit inside the
+    maximum bit length specified by our snowflake ID schema.
+    """
+
+    address: str
+    """The address of the region's silo.
+
+    In a production environment, this would typically be a subdomain (e.g.,
+    "eu.sentry.io"). In a development environment, it may be useful to define
+    region addresses as a custom hostname or port number (e.g,
+    "localhost:8001").
+    """
+
+    category: RegionCategory
+    """The region's category."""
+
+    def __post_init__(self) -> None:
+        from sentry.utils.snowflake import REGION_ID
+
+        REGION_ID.validate(self.id)
+
+    def to_url(self, path: str) -> str:
+        return self.address + path
+
+
+class RegionMapping:
+    """The set of all regions in this Sentry platform instance."""
+
+    def __init__(self, regions: Iterable[Region]) -> None:
+        self.regions = tuple(regions)
+        self._by_name = {r.name: r for r in self.regions}
+        self._by_id = {r.id: r for r in self.regions}
+
+    @classmethod
+    def load_from_config(cls) -> RegionMapping:
+        from django.conf import settings
+
+        # For now, assume that all region configs can be taken in through Django
+        # settings. We may investigate other ways of delivering those configs in
+        # production.
+        return cls(settings.SENTRY_REGION_CONFIG)
+
+    def get_by_name(self, name: str) -> Region | None:
+        """Look up a region by name."""
+        return self._by_name.get(name)
```

Comment:
Is there possibly some merit to explicitly throwing exception with useful error message when lookups fail?  The problem is almost certainly consumers will all either be written to do the same -- check for null and throw a useful error message, or worse, do no checking and result in unuseful NoneType exceptions.

Reply:
Exceptions is better imo. A region being missing and an organization not belonging to a region when the application is expecting a region feel like 'exceptional' scenarios.

Reply:
👍 I was on the fence myself. Switched to exceptions in 03d8911f4335954756269c4e28e3cd74e711707b